packages/gatsby/src/utils/jobs-manager.js

```diff
@@ -0,0 +1,237 @@
+    process.nextTick(() => {
+      try {
+        resolve(
+          workerFn({
+            inputPaths: job.inputPaths,
+            outputDir: job.outputDir,
+            args: job.args,
+          })
+        )
+      } catch (err) {
+        reject(err)
+      }
+    })
+  })
+}
+
+/**
+ *
+ * @param {InternalJob} job
+ */
+const runJob = job => {
+  const { plugin } = job
+  try {
+    const worker = require(path.posix.join(plugin.resolve, `gatsby-worker.js`))
+    if (!worker[job.name]) {
+      throw new Error(`No worker function found for ${job.name}`)
+    }
+
+    return runLocalWorker(worker[job.name], job)
+  } catch (err) {
+    throw new Error(
+      `We couldn't find a gatsby-worker.js(${plugin.resolve}/gatsby-worker.js) file for ${plugin.name}@${plugin.version}`
+    )
+  }
+}
+
+/**
+ * Create an internal job object
+ *
+ * @param {JobInput|InternalJob} job
+ * @param {{name: string, version: string, resolve: string}} plugin
+ * @return {InternalJob}
+ */
+exports.createInternalJob = (job, plugin) => {
+  // It looks like we already have an augmented job so we shouldn't redo this work
+  // @ts-ignore
+  if (job.id && job.contentDigest) {
+    return job
+  }
+
+  const { name, inputPaths, outputDir, args } = job
+
+  // TODO see if we can make this async, filehashing might be expensive to wait for
+  // currently this needs to be sync as we could miss jobs to have been scheduled and
+  // are still processing their hashes
+  const inputPathsWithContentDigest = inputPaths.map(path => {
+    return {
+      path: convertPathsToAbsolute(path),
+      contentDigest: createFileHash(path),
+    }
+  })
+
+  /** @type {InternalJob} */
+  const internalJob = {
+    id: uuid(),
+    name,
+    contentDigest: ``,
+    inputPaths: inputPathsWithContentDigest,
+    outputDir: convertPathsToAbsolute(outputDir),
+    args,
+    plugin: {
+      name: plugin.name,
+      version: plugin.version,
+      resolve: plugin.resolve,
+      isLocal: !plugin.resolve.includes(`/node_modules/`),
+    },
+  }
+
+  internalJob.contentDigest = createContentDigest({
+    name: job.name,
+    inputPaths: internalJob.inputPaths.map(
+      inputPath => inputPath.contentDigest
+    ),
+    outputDir: internalJob.outputDir,
+    args: internalJob.args,
+    plugin: internalJob.plugin,
+  })
+
+  return internalJob
+}
+
+/**
+ * Creates a job
+ *
+ * @param {InternalJob} job
+ * @return {Promise<object>}
+ */
+exports.enqueueJob = async job => {
+  // When we already have a job that's executing, return the same promise.
+  if (jobsInProcess.has(job.contentDigest)) {
+    return jobsInProcess.get(job.contentDigest).deferred.promise
+  }
+
+  // Bump active jobs
+  activeJobs++
+  if (!activityForJobs) {
+    activityForJobs = reporter.phantomActivity(`Running jobs`)
+    activityForJobs.start()
+  }
+
+  const deferred = pDefer()
+  jobsInProcess.set(job.contentDigest, {
+    id: job.id,
+    deferred,
+  })
```

Comment:
Do we ever clean up `jobsInProcess` once job is done? (I couldn't find anything).

Reply:
No, but we don't have to as the map has all executed jobs in the current process so we need it to return the same process but when using redux this becomes moot as it's inside the done state anyway 🙄.